packages/Button/src/Button.tsx

```diff
@@ -267,61 +281,124 @@ const baseStyle = css`
   }
 `;
 
-export default class Button extends Component {
-  static displayName = 'Button';
+interface SharedButtonProps {
+  variant: Variant;
+  size: Size;
+  className?: string;
+  children?: React.ReactNode;
+  disabled?: boolean;
+}
 
-  static propTypes = {
-    variant: PropTypes.oneOf(['default', 'primary', 'info', 'danger', 'dark']),
-    size: PropTypes.oneOf(['xsmall', 'small', 'normal', 'large']),
-    className: PropTypes.string,
-    children: PropTypes.node,
-    disabled: PropTypes.bool,
-    as: PropTypes.oneOfType([PropTypes.string, PropTypes.element]),
-    href: PropTypes.string,
-  };
+interface LinkButtonProps extends HTMLElementProps<'a'>, SharedButtonProps {
+  href: string;
+}
+
+interface ButtonButtonProps
+  extends HTMLElementProps<'button'>,
+    SharedButtonProps {
+  href?: null;
+}
+
+type CustomElementButtonProps = SharedButtonProps & {
+  as: React.ElementType<any>;
+  [key: string]: any;
+};
+
+type ButtonProps =
+  | LinkButtonProps
+  | ButtonButtonProps
+  | CustomElementButtonProps;
+
+function hasCustomRoot(props: ButtonProps): props is CustomElementButtonProps {
+  return (props as any).as != null;
+}
+
+function hasLinkRoot(
+  props: LinkButtonProps | ButtonButtonProps,
+): props is LinkButtonProps {
+  return props.href != null;
+}
+
+const omitProps = [
+  'as',
+  'className',
+  'disabled',
+  'size',
+  'variant',
+  'children',
+] as const;
+
+export default class Button extends Component<ButtonProps> {
+  static displayName = 'Button';
 
   static defaultProps = {
     variant: 'default',
     size: 'normal',
     className: '',
     children: null,
     disabled: false,
-    type: 'button',
   };
 
   render() {
-    const {
-      className,
-      children,
+    const { className, children, disabled, variant, size } = this.props;
+
+    const commonProps = {
+      className: cx(
+        baseStyle,
+        buttonSizes[size],
+        buttonVariants[variant],
+        className,
+      ),
       disabled,
-      variant,
-      size,
-      as,
-      href,
-      ...rest
-    } = this.props;
+      'aria-disabled': disabled,
+    };
 
-    let Root = href ? 'a' : 'button';
+    if (hasCustomRoot(this.props)) {
+      const Root = this.props.as;
+      const rest = omit(this.props, omitProps);
 
-    if (as) {
-      Root = as;
+      return (
+        <Root {...rest} {...commonProps}>
+          {children}
+        </Root>
+      );
     }
 
+    if (hasLinkRoot(this.props)) {
+      const rest: HTMLElementProps<'a'> = omit(this.props, omitProps);
+      return (
+        <a {...rest} {...commonProps}>
+          {children}
+        </a>
+      );
+    }
+
+    const rest: HTMLElementProps<'button'> = omit(this.props, omitProps);
```

Comment:
We have to split this line up into 3 places for type reasons? Boo. We can't just cast it as appropriate? Seems silly to dupe it.

Reply:
Oh, I can do `{...rest as HTMLElementProps<'button'>}`. That seems nicer, I guess. I'll do that.

Reply:
Question: Does the HTMLElementProps type allow for `data-*` and `aria-*` props?

Reply:
I'd expect it to, although I don't know exactly how. But as far as I understand, it's accessing the same types as the core JSX parser with a `<div>` or whatever